### ISSUE TITLE
Tracking issue for Transitioning crates to Rust 2018

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#58099]
This issue tracks the transitioning of all crates in this repo and sub-modules to Rust 2018.

-----------------------------------

You can help by filing PRs transitioning a crate at a time to Rust 2018.
When transitioning, please apply the following to the crate root:
```rust
#![deny(rust_2018_idioms)]
#![allow(explicit_outlives_requirements)] // Remove or keep this at your option.
```

You can try to use `cargo +nightly fix` (or `cargo +nightly --edition-idioms`) to migrate a crate automatically; see the [edition guide](https://rust-lang-nursery.github.io/edition-guide/editions/transitioning-an-existing-project-to-a-new-edition.html) for details. However, due to bootstrapping, you might need to do some manual intervention here and there.

If you want to help out, please pick a crate and leave a comment here.
When you make a transitioning PR, also leave a comment here.

cc @rust-lang/compiler @rust-lang/infra

----------------------------------------------------

The following crates exist in the repo or as submodules.
Checked items have been transitioned to Rust 2018.

- [x] bootstrap
- [x] build_helper; done in https://github.com/rust-lang/rust/pull/58120
- [x] liballoc; done in https://github.com/rust-lang/rust/pull/58081
- [x] libarena; done in https://github.com/rust-lang/rust/pull/58105
- [x] libcore; done in https://github.com/rust-lang/rust/pull/58702
- [x] libfmt_macros; done in https://github.com/rust-lang/rust/pull/58106
- [x] libgraphviz; done in https://github.com/rust-lang/rust/pull/58107
- [x] libpanic_abort; done in https://github.com/rust-lang/rust/pull/58112
- [x] libpanic_unwind; done in https://github.com/rust-lang/rust/pull/58110
- [x] libproc_macro; done in https://github.com/rust-lang/rust/pull/58119
- [x] libprofiler_builtins; done in https://github.com/rust-lang/rust/pull/58220
- [x] librustc; done in https://github.com/rust-lang/rust/pull/58193
- [x] librustc_allocator; done in https://github.com/rust-lang/rust/pull/58222
- [x] librustc_apfloat; done in https://github.com/rust-lang/rust/pull/58239
- [x] librustc_asan; done in https://github.com/rust-lang/rust/pull/58219
- [x] librustc_borrowck; done in https://github.com/rust-lang/rust/pull/58260
- [x] librustc_codegen_llvm; done in https://github.com/rust-lang/rust/pull/58546
- [x] librustc_codegen_ssa; done in https://github.com/rust-lang/rust/pull/58322
- [x] librustc_codegen_utils; done in https://github.com/rust-lang/rust/pull/58259
- [x] librustc_cratesio_shim; done in https://github.com/rust-lang/rust/pull/58256
- [x] librustc_data_structures; done in https://github.com/rust-lang/rust/pull/58312
- [x] librustc_driver; done in https://github.com/rust-lang/rust/pull/58253
- [x] librustc_errors; done https://github.com/rust-lang/rust/pull/58240
- [x] librustc_fs_util; done in https://github.com/rust-lang/rust/pull/58225
- [x] librustc_incremental; done in https://github.com/rust-lang/rust/pull/58261
- [x] librustc_interface; done in https://github.com/rust-lang/rust/pull/58837
- [x] librustc_lint; done in https://github.com/rust-lang/rust/pull/58245
- [x] librustc_llvm; done in https://github.com/rust-lang/rust/pull/58241
- [x] librustc_lsan; done in https://github.com/rust-lang/rust/pull/58223
- [x] librustc_metadata; done in https://github.com/rust-lang/rust/pull/58255
- [x] librustc_mir; done in https://github.com/rust-lang/rust/pull/58265
- [x] librustc_msan; done in https://github.com/rust-lang/rust/pull/58218
- [x] librustc_passes; done in https://github.com/rust-lang/rust/pull/58247
- [x] librustc_plugin; done in https://github.com/rust-lang/rust/pull/58228
- [x] librustc_privacy; done in https://github.com/rust-lang/rust/pull/58109
- [x] librustc_resolve; done in https://github.com/rust-lang/rust/pull/58236
- [x] librustc_save_analysis; done in https://github.com/rust-lang/rust/pull/58233
- [x] librustc_target; done in https://github.com/rust-lang/rust/pull/58257
- [x] librustc_traits; done in https://github.com/rust-lang/rust/pull/58251
- [x] librustc_tsan; done in https://github.com/rust-lang/rust/pull/58217
- [x] librustc_typeck; done in https://github.com/rust-lang/rust/pull/58300
- [x] librustdoc; done in https://github.com/rust-lang/rust/pull/58100
- [x] libserialize; done in https://github.com/rust-lang/rust/pull/58318
- [x] libstd; done in https://github.com/rust-lang/rust/pull/58208
- [x] libsyntax; done in https://github.com/rust-lang/rust/pull/58125
- [x] libsyntax_ext; done in https://github.com/rust-lang/rust/pull/58133
- [x] libsyntax_pos; done in https://github.com/rust-lang/rust/pull/58124
- [x] libterm; done in https://github.com/rust-lang/rust/pull/58111
- [x] libtest; done in https://github.com/rust-lang/rust/pull/58118
- [x] libunwind; done in https://github.com/rust-lang/rust/pull/58128
- [x] rustc
- [x] stdsimd
- tools
    - [x] build-manifest; done in https://github.com/rust-lang/rust/pull/58113
    - [x] cargo - submodule: https://github.com/rust-lang/cargo/
    - [x] cargo-test; done in https://github.com/rust-lang/rust/pull/58095
    - [x] clippy - submodule: https://github.com/rust-lang/rust-clippy
        - [x] clippy_dev
        - [x] clippy_dummy
        - [x] clippy_lints
        - [x] clippy_workspace_tests; fixed in https://github.com/rust-lang/rust-clippy/pull/3737
        - [x] mini-macro; fixed in https://github.com/rust-lang/rust-clippy/pull/3737
        - [x] rustc_tools_util
    - [x] compiletest; done in https://github.com/rust-lang/rust/pull/58091
    - [x] error_index_generator; done in https://github.com/rust-lang/rust/pull/58285
    - [x] linkchecker; done in https://github.com/rust-lang/rust/pull/58096
    - [x] miri - submodule: https://github.com/solson/miri
    - [x] remote-test-client; done in https://github.com/rust-lang/rust/pull/58097
    - [x] remote-test-server; done in https://github.com/rust-lang/rust/pull/58097
    - [x] rls - submodule: https://github.com/rust-lang/rls
    - [x] rust-installer
    - [x] rustbook
    - [x] rustc-std-workspace-core; done in https://github.com/rust-lang/rust/pull/58287
    - [x] rustc-workspace-hack; done in https://github.com/rust-lang/rust/pull/58288
    - [x] rustdoc; done in https://github.com/rust-lang/rust/pull/58115
    - [x] rustdoc-themes; done in https://github.com/rust-lang/rust/pull/58117
    - [x] rustfmt - submodule: https://github.com/rust-lang/rustfmt
    - [x] tidy; done in https://github.com/rust-lang/rust/pull/58114
    - [x] unstable-book-gen; done in https://github.com/rust-lang/rust/pull/58114

**Comment by gnzlbg** [comment]
@Centril  

FWIW `packed_simd`, the old `std::simd` module, is already using Rust 2018. 

> stdsimd - cc @gnzlbg; this blocks the transition of libcore

The challenge is that `libcore` also blocks the transition of `stdsimd` to the 2018 edition because if we update `stdsimd`, we can't update rust-lang/rust with a new stdsimd version again until `libcore` is a Rust2018 crate. 

That is, doing this upgrade is going to require some coordination. If someone wants to work on the `libcore` part of things, it might be a good idea to ping me on Discord / Zulip so that we can talk about the timeline.

Basically, I'll send a PR to stdsimd to update it to the Rust2018 edition but won't merge it. Then somebody can send a PR to rust-lang/rust updating libcore to Rust2018, using the stdsimd branch from my PR. Once that passes all tests, I'll merge the stdsimd PR, and the rust-lang/rust libcore PR can be updated to use stdsimd master, and merged.

**Comment by mark-i-m** [comment]
I would like to take a stab at `librustc`. I had previously played around with it and found it annoying because bootstrap sets some env variables, but I think I might be able to get it this time.

**Comment by mark-i-m** [comment]
Some notes as I'm working on this:

- Compiling librustc requires having some environment variables set. Currently, I'm using
   ```bash
   CFG_COMPILER_HOST_TRIPLE=x86_64-unknown-linux-gnu RUSTC_ERROR_METADATA_DST=/tmp/foo cargo +nightly fix
   ```

- For some reason, I'm unable to compile benchmarks due to weird import errors with `libtest`. I'm assuming that these are caused by bootstrapping. In this case, I'm temporarily deleting the offending files...

- You will need to use the `--broken-code` flag. It turns out that `cargo fix` has _very weird_ interactions with macros. For example, it tried to make `dyn` a raw ident in a macro, rather than leaving it alone, leading to compile errors like this:

```
error: expected one of `!`, `(`, `)`, `,`, `::`, or `<`, found `dep_tracking`
    --> src/librustc/session/config.rs:290:51
     |
290  |                                 $opt_expr as &r#dyn dep_tracking::DepTrackingHash).is_some() {
     |                                                     ^^^^^^^^^^^^ expected one of `!`, `(`, `)`, `,`, `::`, or `<` here
...
1145 | / options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
1146 | |           build_debugging_options, "Z", "debugging",
1147 | |           DB_OPTIONS, db_type_desc, dbsetters,
1148 | |     codegen_backend: Option<String> = (None, parse_opt_string, [TRACKED],
...    |
1401 | |          the same values as the target option of the same name"),
1402 | | }
     | |_- in this macro invocation
```

Instead, I'm going to apply the majority of changes using `--broken-code`. Then, I will go back and fix the problems rustfix introduces... I'm guessing this is a bug, but honestly, I don't even know what to report...

**Comment by cramertj** [comment]
@gnzlbg 
> The challenge is that `libcore` also blocks the transition of `stdsimd` to the 2018 edition because if we update `stdsimd`, we can't update rust-lang/rust with a new stdsimd version again until `libcore` is a Rust2018 crate.

I'd do this by making `stdsimd` edition-agnostic (compile on both editions). Then libcore can update to 2018 edition and `stdsimd` can follow.

**Comment by gnzlbg** [comment]
> I'd do this by making `stdsimd` edition-agnostic (compile on both editions). Then libcore can update to 2018 edition and `stdsimd` can follow.

Great idea!

**Comment by gnzlbg** [comment]
So I've applied `cargo fix` to `core_arch` and to `std_detect` and it reports zero changes, so porting libcore should just work. If someone wants to attempt to port libcore to the Rust2018 edition please ping me if you encounter any issues with `stdsimd`.

**Comment by taiki-e** [comment]
I'm currently working on the transition of `libstd`, but I'd like to work on the transition of `libcore` if that is over (If someone else is not working on it).

**Comment by mark-i-m** [comment]
@Centril It looks like the checklist needs to be updated.

**Comment by mati865** [comment]
`libstd` is merged https://github.com/rust-lang/rust/pull/58208

**Comment by Centril** [comment]
Yep; still waiting on https://github.com/rust-lang/rust/pull/58828 however.

**Comment by petrochenkov** [comment]
It should be possible to set the default edition and lints in rustbuild already.
The few remaining crates can opt-out and be transitioned at later time, not necessary to wait for them.

**Comment by Centril** [comment]
@pietroalbini @Mark-Simulacrum I'm pleased to announce that all crates, to my knowledge, have been migrated to Rust 2018. As far as I know, all of them have also had idiom lints applied. I think now would be a good idea to apply it globally for rustbuild.

**Comment by Mark-Simulacrum** [comment]
I can think of a couple of things we want:
 * tidy-enforced that all new crates are in 2018 edition
 * deny idiom lints

Not entirely sure how best to do the first, but the second should be quite simple -- I would add to https://github.com/rust-lang/rust/blob/master/src/bootstrap/bin/rustc.rs#L307 the deny for idiom lints.

**Comment by mark-i-m** [comment]
This is done now, right? Can we close this?

**Comment by Centril** [comment]
@mark-i-m 

> tidy-enforced that all new crates are in 2018 edition

is not done I believe.

**Comment by crlf0710** [comment]
Finally done! Maybe tick the checkbox too!

**Comment by crlf0710** [comment]
Finally done! Maybe tick the checkbox too!
